apps/remote_control/test/lexical/remote_control/search/store/backends/ets/wal_test.exs

```diff
@@ -0,0 +1,276 @@
+      Wal.close(new_wal)
+      # write junk over it
+      File.write!(checkpoint_path, "this is not data")
+
+      {:ok, new_wal} =
+        wal_state.project
+        |> Wal.new(1, @table_name)
+        |> Wal.load()
+
+      assert Wal.size(new_wal) == {:ok, 0}
+      assert new_wal.checkpoint_version == 0
+    end
+
+    test "can load a checkpoint", %{wal: wal_state} do
+      with_wal wal_state do
+        :ets.insert(@table_name, [{:first, 1}, {:second, 2}])
+      end
+
+      assert wal_state.checkpoint_version == 0
+
+      # prior, we had no checkpoint and one item in the update
+      # log. Checkpointing clears out the updates log and
+      # creates a checkpoint file, which can be restored
+      assert {:ok, 1} = Wal.size(wal_state)
+      assert {:ok, new_wal} = Wal.checkpoint(wal_state)
+
+      checkpoint_version = new_wal.checkpoint_version
+
+      assert checkpoint_version > 0
+      assert {:ok, 0} = Wal.size(new_wal)
+
+      items = dump_and_close_table()
+      {:ok, loaded_wal, ^items} = load_from_wal(new_wal)
+      assert loaded_wal.checkpoint_version == checkpoint_version
+    end
+
+    test "can handle lots of data", %{wal: wal_state} do
+      stream =
+        1..500_000
+        |> Stream.cycle()
+        |> Stream.map(fn count -> {{:item, count}, count} end)
+
+      for item <- Enum.take(stream, 20_000) do
+        with_wal wal_state do
+          :ets.insert(@table_name, item)
+        end
+      end
+
+      {:ok, new_state} = Wal.checkpoint(wal_state)
+      :ok = Wal.close(new_state)
+      data = dump_and_close_table()
+
+      assert {:ok, _wal_state, entries} =
+               new_state.project
+               |> Wal.new(1, @table_name)
+               |> load_from_wal()
+
+      assert Enum.sort(entries) == Enum.sort(data)
+    end
+
+    test "checkpoints after a certain number of operations", %{project: project} do
+      {:ok, wal_state} =
+        project
+        |> Wal.new(1, @table_name, max_wal_operations: 5)
+        |> Wal.load()
+
+      with_wal wal_state do
+        :ets.insert(@table_name, {:first, 1})
+        :ets.insert(@table_name, {:first, 2})
+        :ets.insert(@table_name, {:first, 3})
+        :ets.insert(@table_name, {:first, 4})
+      end
+
+      assert Wal.size(wal_state) == {:ok, 4}
+
+      with_wal wal_state do
+        :ets.insert(@table_name, {:first, 5})
+      end
+
+      assert Wal.size(wal_state) == {:ok, 0}
+    end
+  end
+
+  defp with_a_loaded_wal(%{wal: wal_state}) do
+    {:ok, wal_state} = Wal.load(wal_state)
+    {:ok, wal: wal_state}
+  end
+
+  defp dump_and_close_table do
+    items = :ets.tab2list(@table_name)
+    :ets.delete(@table_name)
+    items
+  end
+
+  defp load_from_wal(wal_state) do
+    new_table()
+    {:ok, new_wal} = Wal.load(wal_state)
+    entries = :ets.tab2list(@table_name)
+    {:ok, new_wal, entries}
+  end
+
+  defp new_table do
+    if :ets.info(@table_name) == :undefined do
+      :ets.new(@table_name, [:named_table, :set])
```

Comment:
Schema v2 uses `:ordered_set`, which I think is necessary for the `:ets.foldl` to traverse in-order. It doesn't seem to matter for these tests, but maybe worth using anyways? Or exposing `Backends.Ets.current_schema()` and using `Backends.Ets.current_schema().table_options()`.